### PR TITLE
expanding ResolutionPolicy for public use

### PR DESCRIPTION
**Pull request by brettsam**
I partially resurrected Mathew's PR here: https://github.com/Azure/azure-webjobs-sdk/pull/971. I'm going to need this to be able to move forward with DocDB so wanted to get an initial review out before writing lots of tests and plumbing i through everything.

One big change that @MikeStall will be interested in -- now passing the ValueBindingContext along with the resolved attribute to all of the builder methods (only ExactType is doing this for this initial PR).

### CODE REVIEW

File: src/Microsoft.Azure.WebJobs/ResolutionContext.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+            _properties = properties;
+        }
+
+        /// <summary>
+        /// The value of the property.
+        /// </summary>
+        public string PropertyValue { get; set; }
+
+        /// <summary>
+        /// Gets or sets the binding data collection.
```
Comment (MikeStall) [inline src/Microsoft.Azure.WebJobs/ResolutionContext.cs]: >or sets [](start = 17, length = 7)

Type is read-only dict. Can data be set?

**Comment by MikeStall** [comment]
>         [AutoResolve]

Did this need a resolution policy marker?  



---
Refers to: src/Microsoft.Azure.WebJobs/TableAttribute.cs:94 in 07f8728. [](commit_id = 07f8728343b137abeb05447e5b4c6b8c170fb2e2, deletion_comment = False)

File: src/Microsoft.Azure.WebJobs/ResolutionContext.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        /// </summary>
+        public IReadOnlyDictionary<string, object> BindingData { get; set; }
+
+        /// <summary>
+        /// Gets or sets property values for the current bind operation.
+        /// </summary>
+        public IDictionary<string, object> Properties
```
Comment (MikeStall) [inline src/Microsoft.Azure.WebJobs/ResolutionContext.cs]: >Properties [](start = 43, length = 10)

What's the difference between properties and binding data?

File: src/Microsoft.Azure.WebJobs.Host/Bindings/AttributeCloner.cs [hunk 3]
```diff
@@ -23,22 +22,26 @@ internal class AttributeCloner<TAttribute>
...
 
+        private IResolutionPolicy _defaultResolutionPolicy;
+
         public AttributeCloner(
             TAttribute source,
             IReadOnlyDictionary<string, Type> bindingDataContract,
             INameResolver nameResolver = null,
-            Func<TAttribute, Task<TAttribute>> hook = null)
+            Func<TAttribute, Task<TAttribute>> hook = null,
+            IResolutionPolicy defaultResolutionPolicy = null)
```
Comment (MikeStall) [inline src/Microsoft.Azure.WebJobs.Host/Bindings/AttributeCloner.cs]: >IResolutionPolicy defaultResolutionPolicy = null [](start = 12, length = 48)

This feels wrong:
1. attribute wide?  What happens when different properties have different resolution policies. 
2. defaulting to null?  Why can't we just grab it off the [AutoResolve] attribute?

Comment (brettsam) [inline src/Microsoft.Azure.WebJobs.Host/Bindings/AttributeCloner.cs]: This is the DefaultResolutionPolicy -- what we fall back to if it isn't specified on the [AutoRenew]. It's not happening yet (I meant to call this out), but this would let someone register their own DefaultResolutionPolicy. Then, they would pull it from the Services and pass it into the BindingFactory like they do with INameResolver today. If we want to allow this (which I believe you called out in the other review), we'd need a way to let someone override the default somewhere.

File: src/Microsoft.Azure.WebJobs/ODataFilterResolutionPolicy.cs [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.Azure.WebJobs
+{
+    /// <summary>
+    /// Marker type for Table resolution
+    /// </summary>
+    public sealed class ODataFilterResolutionPolicyMarker<TAttribute> where TAttribute : Attribute
```
Comment (MikeStall) [inline src/Microsoft.Azure.WebJobs/ODataFilterResolutionPolicy.cs]: >TAttribute [](start = 58, length = 10)

Why is this needed?

File: src/Microsoft.Azure.WebJobs.Host/Bindings/DefaultResolutionPolicy.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+using Microsoft.Azure.WebJobs.Host.Bindings.Path;
+
+namespace Microsoft.Azure.WebJobs
+{
+    /// <summary>
+    /// Resolution policy for { } in  binding templates. 
+    /// The default policy is just a direct substitution for the binding data. 
+    /// Derived policies can enforce formatting / escaping when they do injection. 
+    /// </summary>
+    internal class DefaultResolutionPolicy<TAttribute> : IResolutionPolicy<TAttribute> where TAttribute : Attribute
```
Comment (MikeStall) [inline src/Microsoft.Azure.WebJobs.Host/Bindings/DefaultResolutionPolicy.cs]: ><TAttribute>  [](start = 42, length = 13)

I'd remove this generic param. 99% of guys don't need it and will ignore it, and the 1% case that needs it can cast.

File: test/Microsoft.Azure.WebJobs.Host.UnitTests/PublicSurfaceTests.cs [hunk 6]
```diff
@@ -126,7 +126,9 @@ public void WebJobsPublicSurface_LimitedToSpecificTypes()
                 "StorageAccountAttribute",
                 "DisableAttribute",
                 "TimeoutAttribute",
-                "TraceLevelAttribute"
+                "TraceLevelAttribute",
+                "IResolutionPolicy",
+                "ResolutionContext"
```
Comment (MikeStall) [inline test/Microsoft.Azure.WebJobs.Host.UnitTests/PublicSurfaceTests.cs]: >ResolutionContext [](start = 17, length = 17)

remove? Don't need a new context class anymore

File: src/Microsoft.Azure.WebJobs.Host/Bindings/AttributeCloner.cs [hunk 7]
```diff
@@ -31,6 +30,8 @@ internal class AttributeCloner<TAttribute>
         // Optional hook for post-processing the attribute. This is intended for legacy hack rules.
         private readonly Func<TAttribute, Task<TAttribute>> _hook;
 
+        private IResolutionPolicy<TAttribute> _defaultResolutionPolicy;
```
Comment (MikeStall) [inline src/Microsoft.Azure.WebJobs.Host/Bindings/AttributeCloner.cs]: >private IResolutionPolicy<TAttribute> _defaultResolutionPolicy [](start = 8, length = 62)

we can get rid of this field

File: src/Microsoft.Azure.WebJobs.Host/Bindings/DefaultResolutionPolicy.cs [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    {
+        public string TemplateBind(PropertyInfo propInfo, TAttribute resolvedAttribute, BindingTemplate template, IReadOnlyDictionary<string, object> bindingData)
+        {
+            if (bindingData == null)
+            {
+                return template?.Pattern;
+            }
```
Comment (MikeStall) [inline src/Microsoft.Azure.WebJobs.Host/Bindings/DefaultResolutionPolicy.cs]: I think this null check should be pushed to the caller (rather than to every implementation of TemplateBind).  The current /dev branch has that.

File: test/Microsoft.Azure.WebJobs.Host.UnitTests/PublicSurfaceTests.cs (hunk 6 above)
Comment (brettsam) [inline test/Microsoft.Azure.WebJobs.Host.UnitTests/PublicSurfaceTests.cs]: yeah, i didn't fix those tests -- figured i'd wait till it was all ready (b/c the names kept changing anyway)

File: src/Microsoft.Azure.WebJobs.Host/Bindings/AttributeCloner.cs (hunk 7 above)
Comment (brettsam) [inline src/Microsoft.Azure.WebJobs.Host/Bindings/AttributeCloner.cs]: true -- i'll just create it when i need it

File: src/Microsoft.Azure.WebJobs.Host/Bindings/DefaultResolutionPolicy.cs (hunk 8 above)
Comment (brettsam) [inline src/Microsoft.Azure.WebJobs.Host/Bindings/DefaultResolutionPolicy.cs]: That means that if bindingData is null, my policy is never called. Maybe someone wants to do something special in that case. I didn't like that it restricted something that I'm trying to handle in a custom way.

File: src/Microsoft.Azure.WebJobs/ODataFilterResolutionPolicy.cs (hunk 4 above)
Comment (brettsam) [inline src/Microsoft.Azure.WebJobs/ODataFilterResolutionPolicy.cs]: This is the empty marker class.

File: src/Microsoft.Azure.WebJobs.Host/Bindings/DefaultResolutionPolicy.cs (hunk 5 above)
Comment (brettsam) [inline src/Microsoft.Azure.WebJobs.Host/Bindings/DefaultResolutionPolicy.cs]: Seems so weird not to use it when AttributeCloner<TAttribute> already knows the type. But fine.

File: src/Microsoft.Azure.WebJobs.Host/Bindings/AttributeCloner.cs [hunk 9]
```diff
@@ -305,47 +307,52 @@ internal TAttribute ResolveFromBindings(IReadOnlyDictionary<string, object> bind
...
+            IResolutionPolicy<TAttribute> policy = null;
+
+            if (autoResolveAttribute != null && autoResolveAttribute.ResolutionPolicyType != null)
             {
-                if (propInfo.DeclaringType == typeof(TableAttribute) && propInfo.Name == "Filter")
+                var ctor = autoResolveAttribute.ResolutionPolicyType.GetConstructor(new Type[] { typeof(IResolutionPolicy<TAttribute>) });
+                if (ctor != null)
                 {
-                    return new TableFilterResolutionPolicy();
+                    policy = (IResolutionPolicy<TAttribute>)Activator.CreateInstance(autoResolveAttribute.ResolutionPolicyType, new object[] { _defaultResolutionPolicy });
```
Comment (brettsam) [inline src/Microsoft.Azure.WebJobs.Host/Bindings/AttributeCloner.cs]: I think I'm going to remove this "nesting" feature to keep the change simpler. None of us need it right now and we can always expand it in the future.

File: src/Microsoft.Azure.WebJobs.Host/Bindings/DefaultResolutionPolicy.cs (hunk 8 above)
Comment (MikeStall) [inline src/Microsoft.Azure.WebJobs.Host/Bindings/DefaultResolutionPolicy.cs]: Exactly. Why would somebody do something special in that case? Every chance we give a derived implementation to unnecessarily specialize is just another source of bugs. 

---
In reply to: [99444656](https://github.com/Azure/azure-webjobs-sdk/pull/1000#discussion_r99444656) [](ancestors = 99444656)

File: src/Microsoft.Azure.WebJobs/TableAttribute.cs [hunk 10]
```diff
@@ -91,7 +91,7 @@ public string RowKey
         /// <summary>
         /// Allow arbitrary table filter. RowKey should be null. 
         /// </summary>
-        [AutoResolve]
+        [AutoResolve(ResolutionPolicyType = typeof(ODataFilterResolutionPolicyMarker<TableAttribute>))]
```
Comment (MikeStall) [inline src/Microsoft.Azure.WebJobs/TableAttribute.cs]: ><TableAttribute> [](start = 84, length = 16)

This generic here feels completely out of place. It has no semantic value here; and is only here because it leaked up from the implementation (and the implementation only needs it because of a rare SQL case that no other binding is using)

File: src/Microsoft.Azure.WebJobs.Host/Bindings/AttributeCloner.cs [hunk 11]
```diff
@@ -305,47 +304,44 @@ internal TAttribute ResolveFromBindings(IReadOnlyDictionary<string, object> bind
...
+            var formatterType = autoResolveAttribute.ResolutionPolicyType;
 
-            // Expand on [AutoResolve] property to make this pluggable, 
-            public static ResolutionPolicy GetPolicy(PropertyInfo propInfo)
+            if (formatterType != null)
             {
-                if (propInfo.DeclaringType == typeof(TableAttribute) && propInfo.Name == "Filter")
+                // Special-case Table as there is no way to declare this ResolutionPolicy
+                // and use BindingTemplate in the Core assembly
+                if (formatterType.FullName == typeof(ODataFilterResolutionPolicy).FullName)
```
Comment (MikeStall) [inline src/Microsoft.Azure.WebJobs.Host/Bindings/AttributeCloner.cs]: >FullName [](start = 34, length = 8)

do we need the .FullName? 
just do :
  (formatterType == typeof(ODataFilterResolutionPolicy))

File: src/Microsoft.Azure.WebJobs.Host/Bindings/IResolutionPolicy.cs [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+{
+    /// <summary>
+    /// Resolution policy for { } in  binding templates. 
+    /// The default policy is just a direct substitution for the binding data. 
+    /// Derived policies can enforce formatting / escaping when they do injection. 
+    /// </summary>
+    public interface IResolutionPolicy
+    {
+        /// <summary>
+        /// Binding
```
Comment (MikeStall) [inline src/Microsoft.Azure.WebJobs.Host/Bindings/IResolutionPolicy.cs]: >Binding [](start = 12, length = 7)

nit: fill in comments

File: src/Microsoft.Azure.WebJobs/ODataFilterResolutionPolicy.cs [hunk 13]
```diff
@@ -0,0 +1,12 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+namespace Microsoft.Azure.WebJobs.Host.Bindings
+{
+    /// <summary>
+    /// Marker type for Table resolution
+    /// </summary>
+    internal sealed class ODataFilterResolutionPolicy
```
Comment (MikeStall) [inline src/Microsoft.Azure.WebJobs/ODataFilterResolutionPolicy.cs]: >internal [](start = 4, length = 8)

It'd be fine to make this public so that other attributes could use the same OData resolution logic that Table does .

**Comment by MikeStall** [comment]
:shipit:

File: src/Microsoft.Azure.WebJobs.Host/Bindings/IResolutionPolicy.cs (hunk 12 above)
Comment (brettsam) [inline src/Microsoft.Azure.WebJobs.Host/Bindings/IResolutionPolicy.cs]: I'm going to hit tests and comments in the next PR

File: src/Microsoft.Azure.WebJobs/ODataFilterResolutionPolicy.cs (hunk 13 above)
Comment (brettsam) [inline src/Microsoft.Azure.WebJobs/ODataFilterResolutionPolicy.cs]: This is the "marker" class. I've actually added "Marker" to the end of it to make that clear. I can make the "real" one public, though.

File: src/Microsoft.Azure.WebJobs.Host/Bindings/AttributeCloner.cs (hunk 11 above)
Comment (brettsam) [inline src/Microsoft.Azure.WebJobs.Host/Bindings/AttributeCloner.cs]: I've changed it to `(formatterType == typeof(ODataFilterResolutionPolicyMarker))`. Not sure how I make them use the same type name on both of these without it being confusing.

File: src/Microsoft.Azure.WebJobs.Host/Bindings/DefaultResolutionPolicy.cs (hunk 8 above)
Comment (brettsam) [inline src/Microsoft.Azure.WebJobs.Host/Bindings/DefaultResolutionPolicy.cs]: My philosophy is -- they asked to be called during binding resolution so we should call them and let them decide. 

But since this will be so rarely used, I'm not that concerned about it. I'll change it to check null first.

**Comment by MikeStall** [comment]
This would fix  https://github.com/Azure/azure-webjobs-sdk/issues/990
